### PR TITLE
Return null from Span#getTransaction when transaction ended

### PR DESCRIPTION
**Pull request by felixbarny**
This avoids returning a reset or potentially even a recycled transaction
Switch MockReporter to validate DT schema

### CODE REVIEW

**Comment by felixbarny** [comment]
This does not really work out. If spans are ended before the transaction, the transaction refers to recycled spans. So we would either have to remove them from the span list or come up with something entirely different.

For example, putting all running transactions in a map, removing them on end and wrapping them in a weak reference, in case they are not ended properly. See also https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/internal/recorder/PendingSpans.java

File: apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java [hunk 1]
```diff
@@ -156,16 +156,14 @@ public Transaction withResult(@Nullable String result) {
         return this;
     }
 
-    @Deprecated
     public List<Span> getSpans() {
-        return spans;
+        // ensures visibility; lock is not likely to have contention
+        synchronized (this) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java]: If the access is synchronized, volatile is not needed. Or do you say we could remove the synchronized in favor of volatile? But what about concurrent writes then?

[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/Transaction.java]: I say for the visibility of the `spans` field reference, volatile would be enough and accurate, and if you use it you don't need to synchronize this specific read. This is without regard to reads or writes **made to the list itself** that will remain synchronized.

**Comment by eyalkoren** [comment]
Let's talk :)

**Pull request closed without merging**